R/060_utils.R

```diff
@@ -0,0 +1,348 @@
+#'
+#' @param header_df a data frame containing header cells.
+#' @param header_fill the criteria according to which blank cells will be filled.
+#' @param formats the formats of the workbook associated with the tidyxl data frame.
+
+fill_blanks_in_col_headers <- function(header_df, header_fill, formats) {
+  if (header_fill == "style") {
+    continue <- TRUE
+    
+    while (continue) {
+      sheet_original <- header_df
+      header_df <- header_df %>% unmerge_cells(strict_merging = FALSE)
+      
+      continue <- !identical(sheet_original, header_df)
+    }
+  }
+  if (header_fill == "local_format_id") {
+    continue <- TRUE
+    
+    while (continue) {
+      sheet_original <- header_df
+      header_df <- header_df %>% unmerge_cells(strict_merging = TRUE)
+      
+      continue <- !identical(sheet_original, header_df)
+    }
+  }
+  
+  if (header_fill == "borders") {
+    filled_join <-
+      header_df %>%
+      add_h_border_groups(formats) %>%
+      dplyr::group_by(h_border_group) %>%
+      dplyr::select(row, col, h_border_group, character) %>%
+      dplyr::mutate(value = ifelse(is.na(character), paste3(character, collapse = " _ ") %>%
+                                     stringr::str_remove_all(" _ "), character)) %>%
+      dplyr::ungroup() %>%
+      dplyr::arrange(h_border_group, row, col) %>%
+      dplyr::select(row, col, character = value)
+    
+    header_df <-
+      header_df %>%
+      dplyr::select(-character) %>%
+      dplyr::left_join(filled_join, by = c("row", "col"))
+  }
+  
+  header_df
+}
+
+
+
+#' Unmerge and fill in blanks for headers
+#'
+#' This is an internal function that function ensures that merged cells are unmerged.
+#'
+#' @param header_df a data frame containing header cells.
+#' @param header_fill he criteria according to which blank cells will be filled.
+#' @param formats the formats of the workbook associated with the tidyxl data frame.
+#' @param direction the direction in which headers are filled.
+
+fill_blanks_in_headers <- function(header_df, header_fill, formats, direction) {
+  if (direction %in% c("N", "S", "up", "down")) {
+    header_df <- suppressMessages(fill_blanks_in_col_headers(header_df, header_fill, formats))
+  } else if (direction %in% c("W", "E", "left", "right")) {
+    header_df <- suppressMessages(fill_blanks_in_row_headers(header_df, header_fill, formats))
+  } else {
+    stop("Please check the direction you are providing")
+  }
+
+  header_df
+}
+
+#' Add a set of variables based on a list of formulae
+#'
+#' This is an internal function that takes a data frame and a list of formulae and appends an arbitrary number of columns to the data frame based on the formulae.
+#'
+#' @param df a data frame containing header cells.
+#' @param form_list list of formulae
+#' @param format the formats associated with the workbook containing the header_df cells.
+#'
+
+reduce_mutated <- function(df, form_list, format) {
```

Comment:
My instinct is that dplyr or rlang already provide this functionality, perhaps via `!!!`.

Reply:
This may be complicated by the fact that form_list is a list containing a quosure and a string, with the string representing a predictable variable name. That said, it may be possible to modify the quosure, rather than put it in a list. In that case it is likely that rlang would have a function that does what we need here. 

There was some funky use of rlang going on at the time of making this function!

Reply:
Yeah, I have to relearn rlang every time I use it!

Reply:
What if `form_list` was a named list of quosures, rather than a list of name-quosure pairs?

Reply:
Yes! That seems to be the way to go. Should end up looking something like this:

```
df <- tibble(a = c(1,2,3), b = c(2,4,6))

quo_vars <-  quos(a + 2, b + 3)
names(quo_vars) <-  c("c","d")

df %>% mutate(!!!quo_vars)
```

Reply:
Implemented in 0ef852413efe7b65216ced50fe39d60b39eda253. Note that this function has been designed to take a spreadsheet range (for example, "B5:E6"), a fmt_* function name (for example, fmt_alignment_indent) or an expression (for example, `!is.na(numeric)`). The string and symbol input need some alteration to work with `mutate` as a quosure. This is what `symbol_expressions_to_quosures` and `string_expressions_to_quosures` do.  Some tidying still to do.